test/models/compute_resources/ec2_test.rb

```diff
@@ -1,59 +1,77 @@
-    assert_equal host, as_admin { cr.associated_host(iface) }
-  end
+      should have_one(:key_pair).with_foreign_key('compute_resource_id').
+        dependent(:destroy)
 
-  describe "find_vm_by_uuid" do
-    it "raises RecordNotFound when the vm does not exist" do
-      cr = mock_cr_servers(Foreman::Model::EC2.new, empty_servers)
-      assert_find_by_uuid_raises(ActiveRecord::RecordNotFound, cr)
-    end
+      test "#associated_host matches any NIC" do
+        host = FactoryGirl.create(:host, :ip => '10.0.0.154')
+        cr = FactoryGirl.build(:ec2_cr)
+        iface = mock('iface1', :public_ip_address => '10.0.0.154', :private_ip_address => "10.1.1.1")
+        assert_equal host, as_admin { cr.associated_host(iface) }
+      end
 
-    it "raises RecordNotFound when the compute raises EC2 error" do
-      cr = mock_cr_servers(Foreman::Model::EC2.new, servers_raising_exception(Fog::Compute::AWS::Error))
-      assert_find_by_uuid_raises(ActiveRecord::RecordNotFound, cr)
-    end
-  end
+      describe "find_vm_by_uuid" do
+        it "raises RecordNotFound when the vm does not exist" do
+          cr = mock_cr_servers(Foreman::Model::EC2.new, empty_servers)
+          assert_find_by_uuid_raises(ActiveRecord::RecordNotFound, cr)
+        end
 
-  context "key pairs" do
-    setup do
-      @aws_key_pairs = []
-      3.times do |i|
-        @aws_key_pairs << AWSKeyPair.new("foreman-#{i}", "13:01:73:0#{i}")
+        it "raises RecordNotFound when the compute raises EC2 error" do
+          cr = mock_cr_servers(Foreman::Model::EC2.new, servers_raising_exception(Fog::Compute::AWS::Error))
+          assert_find_by_uuid_raises(ActiveRecord::RecordNotFound, cr)
+        end
       end
-    end
 
-    test "#get_compute_key_pairs" do
-      cr = FactoryGirl.build(:ec2_cr)
-      key_pair = FactoryGirl.build(:key_pair)
-      cr.key_pair = key_pair
-      Foreman::Model::EC2.any_instance.stubs(:key_pairs).returns(@aws_key_pairs)
-      assert_kind_of(ComputeResourceKeyPair, cr.get_compute_key_pairs.first)
-    end
+      context "key pairs" do
+        setup do
+          @aws_key_pairs = []
+          3.times do |i|
+            @aws_key_pairs << AWSKeyPair.new("foreman-#{i}", "13:01:73:0#{i}")
+          end
+        end
 
-    test "should be capable of key_pair" do
-      cr = FactoryGirl.create(:ec2_cr)
-      assert_includes(cr.capabilities, :key_pair)
+        test "#get_compute_key_pairs" do
+          cr = FactoryGirl.build(:ec2_cr)
+          key_pair = FactoryGirl.build(:key_pair)
+          cr.key_pair = key_pair
+          Foreman::Model::EC2.any_instance.stubs(:key_pairs).returns(@aws_key_pairs)
+          assert_kind_of(ComputeResourceKeyPair, cr.get_compute_key_pairs.first)
+        end
+
+        test "should be capable of key_pair" do
+          cr = FactoryGirl.create(:ec2_cr)
+          assert_includes(cr.capabilities, :key_pair)
+        end
+
+        test 'should remote the key pair in the compute resource' do
```

Comment:
1. s/remote/remove/
1. ... on compute resource deletion
1. Should this be in a test for the concern, rather than in the EC2 test? It isn't specific to EC2.
1. What asserts that the key pair is removed?